### PR TITLE
CORE-4328 - threshold selection

### PR DESCRIPTION
**Pull request by amaciaszek-dsai**
From _README.md_:

8. Threshold optimization:
    1. In experiment config in `DetectionMetrics` set `conf: 0.01`, leave `nms_type` as default (_iterative_) and do not set `class_thresholds`
    2. Run `unstructured_sg/scripts/scripts/find_detection_score_threshold_custom.py` script. Update the `THRESHOLD_OPTIMIZATION_FUNCTION` if needed.

### CODE REVIEW

File: unstructured_sg/scripts/find_detection_score_threshold_custom.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+Find the best confidence score threshold for each class in object detection tasks
+Use this script when you have a trained model and want to analyze / optimize its performance
+The thresholds can be used later when performing NMS
+Usage is similar to src/super_gradients/evaluate_from_recipe.py
+
+Notes:
+    This script does NOT run TRAINING, so make sure in the recipe that you load a PRETRAINED MODEL
+    either from one of your checkpoint or from a pretrained model.
+
+General use: python -m super_gradients.scripts.find_detection_score_threshold --config-name="DESIRED_RECIPE" architecture="DESIRED_ARCH"
```
Comment (mariannaparzych) [inline unstructured_sg/scripts/find_detection_score_threshold_custom.py:16]: This "usage" should IMO descire this function, not "the original one from SG"

File: unstructured_sg/scripts/find_detection_score_threshold_custom.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+from super_gradients.common.environment.cfg_utils import add_params_to_cfg, override_cfg
+from super_gradients import Trainer, init_trainer
+from super_gradients.common.object_names import Metrics
+from super_gradients.common.registry import register_metric
+from super_gradients.training.metrics import DetectionMetrics
+from unstructured_sg.utils.metrics import f_score, compute_detection_metrics_custom
+
+
+THRESHOLD_OPTIMIZATION_FUNCTION = partial(f_score, factor=np.sqrt(2))
```
Comment (mariannaparzych) [inline unstructured_sg/scripts/find_detection_score_threshold_custom.py:40]: mybe this should be an arguments with default 2?

File: unstructured_sg/scripts/find_detection_score_threshold_custom.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+        if len(accumulated_matching_info):
+            matching_info_tensors = [torch.cat(x, 0) for x in list(zip(*accumulated_matching_info))]
+
+            # shape (n_class, nb_iou_thresh)
+            (
+                ap_per_present_classes,
+                precision_per_present_classes,
+                recall_per_present_classes,
+                f1_per_present_classes,
```
Comment (mariannaparzych) [inline unstructured_sg/scripts/find_detection_score_threshold_custom.py:75]: why f1, if it is custom f-score?

File: unstructured_sg/utils/metrics.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    preds_scores: torch.Tensor,
+    n_targets: int,
+    recall_thresholds: torch.Tensor,
+    score_threshold: float,
+    device: str,
+    calc_best_score_thresholds=None,
+    threshold_optimization_function: Callable = f_score,
+):
+    """
+    This is a copy of super gradients compute_detection_metrics_per_cls_custom function, but with custom threshold_optimization_function.
```
Comment (mariannaparzych) [inline unstructured_sg/utils/metrics.py:29]:
```suggestion
    This is a copy of super gradients compute_detection_metrics_per_cls function, but with custom threshold_optimization_function.
```

File: unstructured_sg/utils/metrics.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    preds_cls: torch.Tensor,
+    targets_cls: torch.Tensor,
+    device: str,
+    recall_thresholds: Optional[torch.Tensor] = None,
+    score_threshold: Optional[float] = 0.1,
+    calc_best_score_thresholds: bool = None,
+    threshold_optimization_function=f_score,
+) -> Tuple:
+    """
+    This is a copy of super gradients compute_detection_metrics_custom function, but with custom threshold_optimization_function.
```
Comment (mariannaparzych) [inline unstructured_sg/utils/metrics.py:170]:
```suggestion
    This is a copy of super gradients compute_detection_metrics function, but with custom threshold_optimization_function.
```

File: unstructured_sg/utils/metrics.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+            threshold_optimization_function=threshold_optimization_function,
+        )
+        ap[cls_i, :] = cls_ap
+        precision[cls_i, :] = cls_precision
+        recall[cls_i, :] = cls_recall
+
+        custom_metric_per_class_per_threshold[cls_i, :] = cls_custom_metric_per_threshold
+        best_score_threshold_per_cls[cls_i] = cls_best_score_threshold
+
+    f1 = 2 * precision * recall / (precision + recall + 1e-16)
```
Comment (mariannaparzych) [inline unstructured_sg/utils/metrics.py:242]: why f1?

File: unstructured_sg/utils/metrics.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+    mean_custom_metric_across_classes = torch.mean(custom_metric_per_class_per_threshold, dim=0)
+    best_score_threshold = all_score_thresholds[torch.argmax(mean_custom_metric_across_classes)]
+
+    return ap, precision, recall, f1, unique_classes, best_score_threshold, best_score_threshold_per_cls
```
Comment (mariannaparzych) [inline unstructured_sg/utils/metrics.py:247]: why return f1 instead of custom f-score?

**Review by mariannaparzych — CHANGES_REQUESTED** [review]

File: unstructured_sg/scripts/find_detection_score_threshold_custom.py (hunk 2 above)
Comment (amaciaszek-dsai) [reply unstructured_sg/scripts/find_detection_score_threshold_custom.py:40]: What do you mean?

File: unstructured_sg/utils/metrics.py (hunk 6 above)
Comment (amaciaszek-dsai) [reply unstructured_sg/utils/metrics.py:242]: as above

File: unstructured_sg/utils/metrics.py (hunk 7 above)
Comment (amaciaszek-dsai) [reply unstructured_sg/utils/metrics.py:247]: as above

File: unstructured_sg/scripts/find_detection_score_threshold_custom.py (hunk 3 above)
Comment (amaciaszek-dsai) [reply unstructured_sg/scripts/find_detection_score_threshold_custom.py:74]: I'd rather keep that this way as AP, precision, recall and F1 are our standard metrics. Now we can switch F-sqrt(2) into any other metric, it can be e.g. recall and the script will work just fine because threshold optimization metric is separate from these _reported metrics_

File: unstructured_sg/scripts/find_detection_score_threshold_custom.py (hunk 1 above)
Comment (amaciaszek-dsai) [reply unstructured_sg/scripts/find_detection_score_threshold_custom.py:15]: Fix'd

File: unstructured_sg/scripts/find_detection_score_threshold_custom.py (hunk 2 above)
Comment (mariannaparzych) [reply unstructured_sg/scripts/find_detection_score_threshold_custom.py:38]: now, to use different metric than f2, you need to make changes in the script - can't we make it argument?

File: unstructured_sg/scripts/find_detection_score_threshold_custom.py (hunk 3 above)
Comment (mariannaparzych) [reply unstructured_sg/scripts/find_detection_score_threshold_custom.py:73]: So we optimize for eg. f2, bc it better suits our case, but we report f1, to be compatiblle with previous work?

File: unstructured_sg/scripts/find_detection_score_threshold_custom.py (hunk 2 above)
Comment (amaciaszek-dsai) [reply unstructured_sg/scripts/find_detection_score_threshold_custom.py:38]: Hmm... As we use hydra in this script we would need to probably make this `THRESHOLD_OPTIMIZATION_FUNCTION` part of the experiment config to make it happen. Another required change would be to define some dictionary with named metrics:
```
{
"f_sqrt2_score": partial(f_score, factor=np.sqrt(2)),
"f2_score" partial(f_score, factor=2),
...
}
```
so that we can get the function by its name from config. I'm not sure if I like it, it seems to me that just changing the metric as value of this `THRESHOLD_OPTIMIZATION_FUNCTION` will be more straightforward :thinking: 
What do you think?

File: unstructured_sg/scripts/find_detection_score_threshold_custom.py (hunk 3 above)
Comment (amaciaszek-dsai) [reply unstructured_sg/scripts/find_detection_score_threshold_custom.py:73]: Yeah, more or less. I think this script won't be used to generate statistics either way as we use _evaluate_from_recipe.py_ for that

File: unstructured_sg/scripts/find_detection_score_threshold_custom.py (hunk 2 above)
Comment (mariannaparzych) [reply unstructured_sg/scripts/find_detection_score_threshold_custom.py:38]: ok, let's keep it that way

**Review by mariannaparzych — APPROVED** [review]
